libcxx/src/support/win32/excptptr.cpp

```diff
@@ -0,0 +1,541 @@
+            memcpy(_Dest, _Src, _PType->sizeOrOffset);
+
+            if (_PType->properties & CT_IsWinRTHandle) {
+                const auto _PUnknown = *static_cast<IUnknown* const*>(_Src);
+                if (_PUnknown) {
+                    _PUnknown->AddRef();
+                }
+            }
+            return;
+        }
+
+#if _EH_RELATIVE_TYPEINFO
+        const auto _CopyFunc = reinterpret_cast<void*>(_ThrowImageBase + _PType->copyFunction);
+#else // ^^^ _EH_RELATIVE_TYPEINFO / !_EH_RELATIVE_TYPEINFO vvv
+        const auto _CopyFunc = _PType->copyFunction;
+#endif // ^^^ !_EH_RELATIVE_TYPEINFO ^^^
+
+        const auto _Adjusted = __AdjustPointer(const_cast<void*>(_Src), _PType->thisDisplacement);
+        if (_PType->properties & CT_HasVirtualBase) {
+#ifdef _M_CEE_PURE
+            reinterpret_cast<void(__clrcall*)(void*, void*, int)>(_CopyFunc)(_Dest, _Adjusted, 1);
+#else // ^^^ defined(_M_CEE_PURE) / !defined(_M_CEE_PURE) vvv
+            _CallMemberFunction2(_Dest, _CopyFunc, _Adjusted, 1);
+#endif // ^^^ !defined(_M_CEE_PURE) ^^^
+        } else {
+#ifdef _M_CEE_PURE
+            reinterpret_cast<void(__clrcall*)(void*, void*)>(_CopyFunc)(_Dest, _Adjusted);
+#else // ^^^ defined(_M_CEE_PURE) / !defined(_M_CEE_PURE) vvv
+            _CallMemberFunction1(_Dest, _CopyFunc, _Adjusted);
+#endif // ^^^ !defined(_M_CEE_PURE) ^^^
+        }
+    }
+} // unnamed namespace
+
+// All exception_ptr implementations are out-of-line because <memory> depends on <exception>,
+// which means <exception> cannot include <memory> -- and shared_ptr is defined in <memory>.
+// To workaround this, we created a dummy class exception_ptr, which is structurally identical to shared_ptr.
+
+_STD_BEGIN
+struct _Exception_ptr_access {
+    template <class _Ty, class _Ty2>
+    static void _Set_ptr_rep(_Ptr_base<_Ty>& _This, _Ty2* _Px, _Ref_count_base* _Rx) noexcept {
+        _This._Ptr = _Px;
+        _This._Rep = _Rx;
+    }
+};
+_STD_END
+
+static_assert(sizeof(exception_ptr) == sizeof(shared_ptr<const _EXCEPTION_RECORD>)
+                  && alignof(exception_ptr) == alignof(shared_ptr<const _EXCEPTION_RECORD>),
+    "std::exception_ptr and std::shared_ptr<const _EXCEPTION_RECORD> must have the same layout.");
+
+namespace {
+    template <class _StaticEx>
+    class _ExceptionPtr_static final : public _Ref_count_base {
+        // reference count control block for special "never allocates" exceptions like the bad_alloc or bad_exception
+        // exception_ptrs
+    private:
+        void _Destroy() noexcept override {
+            // intentionally does nothing
+        }
+
+        void _Delete_this() noexcept override {
+            // intentionally does nothing
+        }
+
+    public:
+        // constexpr, TRANSITION P1064
+        explicit _ExceptionPtr_static() noexcept : _Ref_count_base() {
+            _PopulateCppExceptionRecord(_ExRecord, &_Ex, static_cast<ThrowInfo*>(__GetExceptionInfo(_Ex)));
+        }
```

Comment:
It seems that we can never make this ctor `constexpr` as `_PopulateCppExceptionRecord` needs to perform `reinterpret_cast` and `RtlPcToFileHeader` (even if `__GetExceptionInfo` gets `constexpr`-ized one day).
Should we drop this comment (also in MSVC STL)?